feat(groupByProp): add type-safe grouping function with tests

bmthd opened the pull request:
Fixes #1057 

The concern is that the implementation type definitions are loose.
Any advice on direction would be appreciated.

---

Make sure that you:

- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`
- [x] New methods added to `docs/src/content/mapping` for both Lodash and Ramda.

---

<details><summary>We use semantic PR titles to automate the release process!</summary>

https://conventionalcommits.org

PRs should be titled following using the format: `< TYPE >(< scope >)?: description`

### Available Types:

- `feat`: new functions, and changes to a function's type that would impact users.
- `fix`: changes to the runtime behavior of an existing function, or refinements to it's type that shouldn't impact most users.
- `test`: tests-only changes (transparent to users of the function).
- `docs`: changes to the documentation of a function **or the documentation site**.
- `build`, `ci`, `chore`, and `revert`: are only relevant for the internals of the library.

For scope put the name of the function you are working on (either new or
existing).

</details>

cjquines commented on packages/remeda/src/groupBy.test-d.ts:
```diff
@@ -124,3 +124,34 @@ describe("filtering on undefined grouper result", () => {
   });
 });
+
+test("infer type of groupBy", () => {
```
**suggestion:** let's put this under its own `describe` block, for static groupby specifically

and let's call this test `infers type of groups`

cjquines replied:
i'd like to add more type tests to the static groupby as well, but i'm not sure what. what do you think?

cjquines commented on packages/remeda/src/groupBy.test.ts:
```diff
@@ -3,7 +3,28 @@ import { pipe } from "./pipe";
 
 describe("data first", () => {
-  test("groupBy", () => {
+  test("groupBy static", () => {
```
**suggestion:** let's make a separate describe block for the static version, as it seems different enough

cjquines reviewed (commented):
**thought:** should this belong to the same function? it has a different signature...

because of how we generate our docs, as well, this wouldn't appear in the docs site with the `key` signature

maybe we should call this `keyBy` and make this a new function, i'm not sure. i'd get a second opinion here. @eranhirsch maybe

eranhirsch commented:
I also think this might make more sense using a different name. Unlike Lodash/Ramda, we prefer functions to be differentiated mainly by name and not to provide overloaded variants with considerably different signatures.

As for the name, I think `groupByProp(data, propName)` makes sense because it replaces the idiomatic `groupBy(data, prop(propName))`. _groupByKey_ seems wrong because the key refers to the prop's (`[name, value]`) name part, but here the value is meaningful to the logic. If we want to be lavish and diverge from conventions, we can call this `pivot` or `pivotOn` e.g,. https://en.wikipedia.org/wiki/Pivot_table. Names are hard.

eranhirsch reviewed (changes requested):
Let's split this out of groupBy.

bmthd commented:
Thank you both for your feedback!
I also think it would be more natural to extract it into a separate function.
If there is no need to force commonality, I think it would be possible to infer the type as a stable tuple type rather than an indeterminate array, so please let me reimplement it.
I will contact you again once I have implemented it.

bmthd commented:
Thank you for your patience.
We have implemented groupByProp, which has been separated as a separate function.
The changes from the previous implementation are as follows
- The type to be received is now limited to Record.
- Split by whether or not it is a readonly Record, and separate static types from non-static types.
- For static types, all arrays are fully inferred by covering arrays.

I think there are still some tests that I have not taken into account, so if you have any suggestions, please let me know.

eranhirsch commented:
Hey, I just wanted to ping this PR to say that I'm really sorry I'm not getting to review it. It's been a really busy month. I hope to get to this next week.

eranhirsch commented:
Hey @bmthd! I'm sorry this got delayed so long. I had a baby in February and haven't had time to review this PR since then. Rebasing this on the latest main is showing this PR is failing type tests for TypeScript 5.1 and 5.2. I want to get this PR shipped because it's an important and valuable addition to the library. So, are you willing to continue working on this together through review(s), or do you prefer I finish the work by myself? I know people are busy, and getting back to notifications on GitHub isn't always practical, so I'll assume that you aren't interested if you don't reply in a week.

Again, I'm sorry that this PR wasn't handled as quickly as I aimed for contributions to the library, and I totally understand if your focus has shifted elsewhere.

bmthd commented:
@eranhirsch 
Hi, thanks for getting back to me and congratulations on the new addition to your family!

I’d love to continue contributing to this, but to be honest, I’m not sure what’s causing the type failures on TypeScript 5.1 and 5.2. I’m happy to try digging into it, but if you already have an idea of the problem or if it’s more efficient for you to take over, I’d be totally fine with that too.

Let me know what you think would work best!

eranhirsch commented:
I have already started chipping away at the problem; I'm sending a few PRs that are related to this type but could be split out. Once those are shipped, I'll push an update to this PR, which should simplify it and also fix the TypeScript issues. You can help me out with this PR by reviewing it, and we'll ship it together.

eranhirsch commented on packages/remeda/src/groupByProp.ts:
```diff
@@ -0,0 +1,188 @@
+  // would work on this object and it would act differently than expected.
+  Object.setPrototypeOf(output, Object.prototype);
+
+  // @ts-expect-error [ts2322] -- This is fine! We use a broader type for output while we build it because it more accurately represents the shape of the object *while it is being built*. TypeScript can't tell that we finished building the object so can't ensure that output matches the expected output at this point.
+  return output;
```
We prefer suppressing ts errors instead of casting so that if future versions of TypeScript infer these the way we want, we can remove the suppressions because suppression for non-errors are errors themselves. Casts often hide type errors and might not surface changes in the typing as effectively.

eranhirsch commented:
Hey @bmthd, Thanks again for all your work getting this utility added. The final code is changed a lot from your version, but it was your version that got me thinking about how we handle other types as well; mainly the idea that we could filter an array by looking at `Record<Prop, Value>` sparked the changes in #1087, and the usage of NonEmptyArray (and @cjquines's #863) drove me to do #1085. In addition, #1080 had some perf optimizations to the runtime implementation that needed to be ported back to this.

I'm really excited about this function because it does a lot of really cool things in a really tight package that would be useful to a lot of projects.

Please feel free to comment, ask questions, and review the PR.

cjquines commented on packages/remeda/src/groupByProp.test.ts:
**question:** can we add a test against #1046 ?

eranhirsch replied:
done!

cjquines reviewed (dismissed):
i would revise the groupByProp documentation if i had more time, but it's not blocking

TkDodo commented:
:tada: This PR is included in version 2.22.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/remeda/v/2.22.0)
- [GitHub release](https://github.com/remeda/remeda/releases/tag/v2.22.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
